webnn/resources/utils.js

```diff
@@ -1,61 +1,312 @@
+    distance = distance >= 0 ? distance : -distance;
+    assert_true(distance <= nulp,
+                `assert_array_approx_equals_ulp: ${description} actual ${actual[i]} should be close enough to expected ${expected[i]} by the acceptable ${nulp} ULP distance, but they have ${distance} ULP distance`);
+  }
+};
+
+/**
+ * Assert actual results with expected results.
+ * @param {String} name - An operation name
+ * @param {(Number[]|Number)} actual
+ * @param {(Number[]|Number)} expected
+ * @param {Number} tolerance
+ * @param {String} operandType  - An operand type string, value: "float32",
+ *     more types, please see:
+ *     https://webmachinelearning.github.io/webnn/#enumdef-mloperandtype
+ * @param {String} metricType - Value: 'ULP', 'ATOL'
+ */
+const doAssert = (name, actual, expected, tolerance, operandType, metricType) => {
+  const description = `test ${name} ${operandType}`;
+  if (typeof expected === 'number') {
+    // for checking a scalar output by matmul 1D x 1D
+    expected = [expected];
+    actual = [actual];
+  }
+  if (metricType === 'ULP') {
+    assert_array_approx_equals_ulp(actual, expected, tolerance, operandType, description);
+  } else if (metricType === 'ATOL') {
+    assert_array_approx_equals(actual, expected, tolerance, description);
+  }
+};
+
+/**
+ * Check computed results with expected data.
+ * @param {String} name - An operation name
+ * @param {Object.<String, MLOperand>} namedOutputOperands
+ * @param {Object.<MLNamedArrayBufferViews>} outputs - The resources of required outputs
+ * @param {Object} resources - Resources used for building a graph
+ */
+const checkResults = (name, namedOutputOperands, outputs, resources) => {
+  const metricType = PrecisionMetrics[name] ? Object.keys(PrecisionMetrics[name])[0] : 'ULP';
+  const tolerance = getPrecisonTolerance(name, metricType, resources);
+  const expected = resources.expected;
+  const operandType = expected.type;
+  let outputData;
+  let expectedData;
+  if (Array.isArray(expected)) {
+    // check outputs by split or gru
+    for (let operandName in namedOutputOperands) {
+      outputData = outputs[operandName];
+      expectedData = getExpectedData(expected, operandName);
+      doAssert(name, outputData, expectedData, tolerance, operandType, metricType)
+    }
+  } else {
+    const outputName = expected.name ? expected.name : 'output';
+    outputData = outputs[outputName];
+    expectedData = expected.data;
+    doAssert(name, outputData, expectedData, tolerance, operandType, metricType)
+  }
+};
+
+/**
+ * Build a graph.
+ * @param {MLGraphBuilder} builder - A ML graph builder
+ * @param {Object} resources - Resources used for building a graph
+ * @param {Function} buildFunc - A build function for an operation
+ * @param {String} name - An operation name
+ * @returns [namedOperands, inputs, outputs]
+ */
+const buildGraph = (builder, resources, buildFunc, name) => {
+  const outputOperand = buildFunc(builder, resources, name);
```

Comment:
I don't find the parameter `name` is used by this method `buildGraph()` and `buildFunc()`. It probably could be removed.

Reply:
Removed it firstly, add this parameter later in next PR, thanks.